qa/tasks/cephfs/test_exports.py

```diff
@@ -627,3 +627,372 @@ def test_ephemeral_pin_shrink_mds(self):
+
+        self.mount_a.setfattr("1", "ceph.dir.pin.random", "1.0")
+        # timing issue
+        # Sometimes, there might be a delay in the synchronization of export_ephemeral_random_pin.
+        time.sleep(10)
+        for i in range(count):
+            self.mount_a.run_shell_payload(f"mkdir -p 1/{i}")
+            self.mount_a.create_n_files(f"1/{i}/file", 100, sync=True)
+
+        self._wait_random_subtrees(count, status=self.status, rank="all")
+
+        self.mount_a.setfattr("1", "ceph.dir.bal.mask", "1")
+        for i in range(count):
+            self.mount_a.create_n_files("1/file", 100, sync=True)
+        self._wait_random_subtrees(count, status=self.status, rank="all")
+
+    def test_unset_rank_mask(self):
+        """
+        That ceph.dir.bal.mask is unset with -1.
+        """
+
+        self.mount_a.setfattr("1", "ceph.dir.bal.mask", "2")
+        value = self.mount_a.getfattr("1", "ceph.dir.bal.mask")
+        self.assertEqual(value, "2")
+        self._wait_subtrees([('/1', 2)], status=self.status)
+
+        self.mount_a.setfattr("1", "ceph.dir.bal.mask", "-1")
+        value = self.mount_a.getfattr("1", "ceph.dir.bal.mask")
+        self.assertEqual(value, None)
+
+    def test_unset_rank_mask_under_nested_root(self):
+        """
+        That ceph.dir.bal.mask is unset under root directory.
+        """
+
+        self.mount_a.setfattr(".", "ceph.dir.bal.mask", "0")
+        self._wait_subtrees([('', 0)], status=self.status, path='', exclude_path='~mds')
+
+        self.mount_a.setfattr("1", "ceph.dir.bal.mask", "2")
+        value = self.mount_a.getfattr("1", "ceph.dir.bal.mask")
+        self.assertEqual(value, "2")
+        self._wait_subtrees([('/1', 2)], status=self.status)
+
+        self.mount_a.setfattr("1", "ceph.dir.bal.mask", "-1")
+        value = self.mount_a.getfattr("1", "ceph.dir.bal.mask")
+        self.assertEqual(value, None)
+
+        self.mount_a.setfattr(".", "ceph.dir.bal.mask", "0")
+        self._wait_subtrees([('', 0)], status=self.status, path='', exclude_path='~mds')
+
+    def test_unset_rank_mask_under_nested_parent(self):
+        """
+        That ceph.dir.bal.mask is unset under nested directory.
+        """
+
+        self.mount_a.setfattr("1", "ceph.dir.bal.mask", "1")
+        self._wait_subtrees([('/1', 1)], status=self.status, rank=1)
+        self.mount_a.setfattr("1/a", "ceph.dir.bal.mask", "2")
+        self._wait_subtrees([('/1', 1), ('/1/a', 2)], status=self.status, rank=2)
+
+        self.mount_a.setfattr("1/a", "ceph.dir.bal.mask", "-1")
+        self._wait_subtrees([('/1', 1)], status=self.status, rank=1)
+
+    def test_mds_failover(self):
+        """
+        That MDS failover does not affect the ceph.dir.bal.mask.
+        """
+        self.config_set('mds', 'mds_debug_subtrees', False)
+
+        self.mount_a.setfattr("1", "ceph.dir.bal.mask", "2")
+        value = self.mount_a.getfattr("1", "ceph.dir.bal.mask")
+        self.assertEqual(value, "2")
```

Comment:
The assert check fails - https://pulpito.ceph.com/vshankar-2024-03-07_10:41:12-fs-wip-vshankar-testing-20240307.022905-testing-default-smithi/7584618/

```
2024-03-07T22:51:06.010 INFO:tasks.cephfs_test_runner:======================================================================
2024-03-07T22:51:06.010 INFO:tasks.cephfs_test_runner:FAIL: test_mds_failover (tasks.cephfs.test_exports.TestBalMask)
2024-03-07T22:51:06.010 INFO:tasks.cephfs_test_runner:That MDS failover does not affect the ceph.dir.bal.mask.
2024-03-07T22:51:06.010 INFO:tasks.cephfs_test_runner:----------------------------------------------------------------------
2024-03-07T22:51:06.011 INFO:tasks.cephfs_test_runner:Traceback (most recent call last):
2024-03-07T22:51:06.011 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_ceph-c_b0ea271bb96f7f4cec2f2a40ff88a5def1e0530f/qa/tasks/cephfs/test_exports.py", line 934, in test_mds_failover
2024-03-07T22:51:06.011 INFO:tasks.cephfs_test_runner:    self.assertEqual(value, "2")
2024-03-07T22:51:06.011 INFO:tasks.cephfs_test_runner:AssertionError: None != '2'
```

Reply:
@vshankar  I will check and get back to you.

Reply:
In the vstart-based local test, all 20 test cases passed. I will look at the log you provided and see if I can reproduce it.

```
2024-03-13 11:38:19,031.031 INFO:__main__:test_unset_rank_mask_under_nested_parent (tasks.cephfs.test_exports.TestBalMask) ... ok
2024-03-13 11:38:19,031.031 INFO:__main__:test_unset_rank_mask_under_nested_root (tasks.cephfs.test_exports.TestBalMask) ... ok
2024-03-13 11:38:19,031.031 INFO:__main__:
2024-03-13 11:38:19,031.031 INFO:__main__:----------------------------------------------------------------------
2024-03-13 11:38:19,031.031 INFO:__main__:Ran 20 tests in 1378.896s
2024-03-13 11:38:19,031.031 INFO:__main__:
2024-03-13 11:38:19,032.032 INFO:__main__:
2024-03-13 11:38:19,032.032 INFO:__main__:> ip netns list
2024-03-13 11:38:19,039.039 INFO:__main__:> sudo ip link delete ceph-brx
Cannot find device "ceph-brx"
2024-03-13 11:38:19,066.066 INFO:__main__:OK
2024-03-13 11:38:19,067.067 INFO:__main__:
```

Reply:
@vshankar 
I think it's a timing issue. This is the case where getfattr fails after setfattr regardless of MDS fail. Can MDS logs be shared?

Reply:
The major difference is the use of the kernel client. I only tested ceph-fuse.

Reply:
The setxattr client request was processed and a getxattr request was requested between early_reply and reply_client_request. I think the client inode xattrs was not updated at that time. Is it possible to guarantee that the client inode's xattrs will be updated after early_reply?

```
2024-03-07T22:50:52.520+0000 7f8607400640  4 mds.0.server handle_client_request client_request(client.5210:27 setxattr #0x10000000000 ceph.dir.bal.mask 2024-03-07T22:50:52.519761+0000 caller_uid=1000, caller_gid=1301{6,36,1000,1301,}) v6
2024-03-07T22:50:52.521+0000 7f8607400640 10 mds.0.server setxattr 'ceph.dir.bal.mask' len 1 on [inode 0x10000000000 [...2,head] /1/ auth v4 ap=2 DIRTYPARENT f(v0 m2024-03-07T22:50:51.975767+0000 1=0+1) n(v0 rc2024-03-07T22:50:51.975767+0000 2=0+2) (isnap sync r=1) (inest lock) (ifile excl) (ixattr xlock x=1 by 0x55c05e997a80) (iversion lock w=1 last_client=5210) caps={5210=pAsLsFsx/pAsLsXsFsxcral@2},l=5210 | request=1 lock=3 dirfrag=1 caps=1 dirtyparent=1 dirty=1 authpin=1 0x55c05e90a100]
2024-03-07T22:50:52.521+0000 7f8607400640 10 mds.0.server early_reply 0 ((0) Success) client_request(client.5210:27 setxattr #0x10000000000 ceph.dir.bal.mask 2024-03-07T22:50:52.519761+0000 caller_uid=1000, caller_gid=1301{6,36,1000,1301,}) v6


2024-03-07T22:50:52.580+0000 7f8607400640  4 mds.0.server handle_client_request client_request(client.5210:28 ??? #0x10000000000 ceph.dir.bal.mask 2024-03-07T22:50:52.578761+0000 caller_uid=1000, caller_gid=1301{6,36,1000,1301,}) v6
2024-03-07T22:50:52.580+0000 7f8607400640  7 mds.0.server reply_client_request -61 ((61) No data available) client_request(client.5210:28 ??? #0x10000000000 ceph.dir.bal.mask 2024-03-07T22:50:52.578761+0000 caller_uid=1000, caller_gid=1301{6,36,1000,1301,}) v6

inode->xattrs may be not updated.
https://github.com/ceph/ceph/blob/main/src/client/Client.cc#L13757

2024-03-07T22:50:55.379+0000 7f85ff200640  7 mds.0.server reply_client_request 0 ((0) Success) client_request(client.5210:27 setxattr #0x10000000000 ceph.dir.bal.mask 2024-03-07T22:50:52.519761+0000 caller_uid=1000, caller_gid=1301{6,36,1000,1301,}) v6
```

Reply:
> The setxattr client request was processed and a getxattr request was requested between early_reply and reply_client_request. I think the client inode xattrs was not updated at that time. Is it possible to guarantee that the client inode's xattrs will be updated after early_reply? Otherwise, should I just add some logic in the test case to wait for a few seconds?

If the client has the relevant caps, it should have the latest xattr version. I think you might have run into a race where the client thinks it has the caps and can trust the cached xattr, however, the xattr is stale.

I have seen something similar before and fixed it in a code path. See commit 526e4ee5b4b7c27306b70de62daa5cf705763130

> 
> ```
> 2024-03-07T22:50:52.520+0000 7f8607400640  4 mds.0.server handle_client_request client_request(client.5210:27 setxattr #0x10000000000 ceph.dir.bal.mask 2024-03-07T22:50:52.519761+0000 caller_uid=1000, caller_gid=1301{6,36,1000,1301,}) v6
> 2024-03-07T22:50:52.521+0000 7f8607400640 10 mds.0.server setxattr 'ceph.dir.bal.mask' len 1 on [inode 0x10000000000 [...2,head] /1/ auth v4 ap=2 DIRTYPARENT f(v0 m2024-03-07T22:50:51.975767+0000 1=0+1) n(v0 rc2024-03-07T22:50:51.975767+0000 2=0+2) (isnap sync r=1) (inest lock) (ifile excl) (ixattr xlock x=1 by 0x55c05e997a80) (iversion lock w=1 last_client=5210) caps={5210=pAsLsFsx/pAsLsXsFsxcral@2},l=5210 | request=1 lock=3 dirfrag=1 caps=1 dirtyparent=1 dirty=1 authpin=1 0x55c05e90a100]
> 2024-03-07T22:50:52.521+0000 7f8607400640 10 mds.0.server early_reply 0 ((0) Success) client_request(client.5210:27 setxattr #0x10000000000 ceph.dir.bal.mask 2024-03-07T22:50:52.519761+0000 caller_uid=1000, caller_gid=1301{6,36,1000,1301,}) v6
> 
> 
> 2024-03-07T22:50:52.580+0000 7f8607400640  4 mds.0.server handle_client_request client_request(client.5210:28 ??? #0x10000000000 ceph.dir.bal.mask 2024-03-07T22:50:52.578761+0000 caller_uid=1000, caller_gid=1301{6,36,1000,1301,}) v6
> 2024-03-07T22:50:52.580+0000 7f8607400640  7 mds.0.server reply_client_request -61 ((61) No data available) client_request(client.5210:28 ??? #0x10000000000 ceph.dir.bal.mask 2024-03-07T22:50:52.578761+0000 caller_uid=1000, caller_gid=1301{6,36,1000,1301,}) v6
> 
> inode->xattrs may be not updated.
> https://github.com/ceph/ceph/blob/main/src/client/Client.cc#L13757
> 
> 2024-03-07T22:50:55.379+0000 7f85ff200640  7 mds.0.server reply_client_request 0 ((0) Success) client_request(client.5210:27 setxattr #0x10000000000 ceph.dir.bal.mask 2024-03-07T22:50:52.519761+0000 caller_uid=1000, caller_gid=1301{6,36,1000,1301,}) v6
> ```